lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx

```diff
@@ -0,0 +1,460 @@
+  PValuesPermutation_ = Point();
+}
+
+/* Get permutation size */
+UnsignedInteger HSICEstimatorImplementation::getPermutationSize() const
+{
+  return permutationSize_;
+}
+
+/* Compute p-value with permutation */
+void HSICEstimatorImplementation::computePValuesPermutation() const
+{
+  const SquareMatrix Wobs(computeWeightMatrix(outputSample_));
+  PValuesPermutation_ = Point(inputDimension_);
+
+  for(UnsignedInteger dim = 0; dim < inputDimension_; ++dim)
+  {
+    const Sample xdim(inputSample_.getMarginal(dim));
+    const Scalar HSIC_obs = computeHSICIndex(xdim, outputSample_, covarianceList_[dim], covarianceList_[inputDimension_], Wobs);
+
+    UnsignedInteger count = 0;
+    for( UnsignedInteger b = 0; b < permutationSize_; ++b)
+    {
+      const Sample Yp(shuffledCopy(outputSample_));
+      const SquareMatrix W(computeWeightMatrix(Yp));
+      const Scalar HSIC_loc = computeHSICIndex(xdim, Yp, covarianceList_[dim], covarianceList_[inputDimension_], W);
+      if( HSIC_loc > HSIC_obs) count += 1;
+    }
+
+    /* p-value by permutation */
+    PValuesPermutation_[dim] = count * 1.0 / (permutationSize_ + 1) ;
+  }
+  isAlreadyComputedPValuesPermutation_ = true ;
+}
+
+/* Get the HSIC indices */
+Point HSICEstimatorImplementation::getHSICIndices() const
+{
+  if(!(isAlreadyComputedIndices_))
+  {
+    computeIndices();
+    isAlreadyComputedIndices_ = true ; 
+  }
+  return HSIC_XY_;
+}
+
+/* Get the R2-HSIC indices */
+Point HSICEstimatorImplementation::getR2HSICIndices() const
+{
+  if(!(isAlreadyComputedIndices_))
+  {
+    computeIndices();
+    isAlreadyComputedIndices_ = true ;
+  }
+  return R2HSICIndices_;
+}
+
+/* Get the p-values by permutation */
+Point HSICEstimatorImplementation::getPValuesPermutation() const
+{
+  if( !(isAlreadyComputedPValuesPermutation_))
+  {
+    computePValuesPermutation();
+    isAlreadyComputedPValuesPermutation_ = true ;
+  }
+  return PValuesPermutation_;
+}
+
+/* Draw the HSIC indices */
+Graph HSICEstimatorImplementation::drawValues(const Point &values, const String &title) const
+{
+
+  if (values.getDimension() == 0) throw InvalidArgumentException(HERE) << "Error: cannot draw cloud based on empty data.";
+  Graph graph(title, "Input marginal number", "", true, "");
+
+  /* Define cloud */
+  Sample data(values.getDimension(), 2);
+  for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
+  {
+    data(k, 0) = k + 1.0;
+    data(k, 1) = values[k];
+  }
+
+  Cloud cloud(data, "red", "circle", "");
+  graph.add(cloud);
+
+  /* Add text description */
+  for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
+  {
+    data(k, 0) = (k + 1.0) + 0.08;
+    data(k, 1) = values[k];
+  }
+
+  Description names(values.getDimension());
+  for(UnsignedInteger i = 0; i < values.getDimension(); ++i)
+  {
+    OSS oss;
+    oss << "X" << i + 1;
+    names[i] = String(oss);
+  }
+
+  Text text(data, names, "right");
+  text.setColor("black");
+  graph.add(text);
+  graph.setXMargin(0.9);
```

Comment:
This produces: 
![image](https://user-images.githubusercontent.com/31351465/150564775-3e446086-a0f5-4b31-a110-380799f8c7e7.png)
which cannot be used. This is because the margins are much too large on the left (there is no negative index), on the right (there is no index greater than 3), and on the bottom (there is no negative HSIC or P-value). For the bottom value, the mathematical HSIC is nonnegative, but its estimate can. 
The correct way to set this is as follows:
```cpp
const UnsignedInteger dimension(values.getDimension());
Scalar maximum_value;
for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
  {
    if (value[i] > maximum_value)
        {
            maximum_value = value[i];
        }
  }
maximum_value_delta = GetResourceMap("HSICEstimator-GraphicsMarginUpperDelta");  # TODO : fill this code correctly and set the default to 0.05
minimum_value_delta = GetResourceMap("HSICEstimator-GraphicsMarginLowerDelta");  # TODO : fill this code correctly and set the default to -0.05
maximum_value  = maximum_value + maximum_value_delta;
lower = [minimum_value_delta, 0.5];  # lower bound of the 2D window
upper = [dimension + 0.5, maximum_value];  # upper bound of the 2D window
Interval boundingBox((lower, upper);
graph.setBoundingBox(boundingBox);
```

Reply:
In the specifications, the following graphics are presented.

![image](https://user-images.githubusercontent.com/31351465/152328095-2f1c3911-a01d-4355-8c01-78ac59e1f51e.png)

![image](https://user-images.githubusercontent.com/31351465/152328180-29046b9f-8937-4acb-9c36-901133903717.png)

To get this result, using setBoundingBox() seems to be the easiest way to me.